yt/utilities/decompose.py

```diff
@@ -103,25 +103,46 @@ def get_psize(n_d, pieces):
     return p_size
 
 
-def split_array(gle, gre, shape, psize):
+def split_array(gle, gre, shape, psize, *, cell_widths=None):
     """Split array into px*py*pz subarrays."""
     n_d = np.array(shape, dtype=np.int64)
     dds = (gre - gle) / shape
     left_edges = []
     right_edges = []
     shapes = []
     slices = []
+
+    if cell_widths is None:
+        cell_widths_by_grid = None
+    else:
+        cell_widths_by_grid = []
+
     for i in range(psize[0]):
         for j in range(psize[1]):
             for k in range(psize[2]):
                 piece = np.array((i, j, k), dtype=np.int64)
                 lei = n_d * piece // psize
                 rei = n_d * (piece + np.ones(3, dtype=np.int64)) // psize
-                lle = gle + lei * dds
-                lre = gle + rei * dds
+
+                if cell_widths is not None:
+                    cws = []
+                    offset_le = []
+                    offset_re = []
+                    for idim in range(3):
+                        cws.append(cell_widths[idim][lei[idim] : rei[idim]])
+                        offset_le.append(np.sum(cell_widths[idim][0 : lei[idim]]))
+                        offset_re.append(offset_le[idim] + np.sum(cws[idim]))
+                    cell_widths_by_grid.append(cws)
+                    offset_re = np.array(offset_re)
+                    offset_le = np.array(offset_le)
+                else:
+                    offset_le = lei * dds
+                    offset_re = rei * dds
+                lle = gle + offset_le
+                lre = gle + offset_re
                 left_edges.append(lle)
                 right_edges.append(lre)
                 shapes.append(rei - lei)
                 slices.append(np.s_[lei[0] : rei[0], lei[1] : rei[1], lei[2] : rei[2]])
 
-    return left_edges, right_edges, shapes, slices
+    return left_edges, right_edges, shapes, slices, cell_widths_by_grid
```

Comment:
this looks like a backward incompatible change. Returning `cell_widths_by_grid` should be optional, and the default behaviour should remain stable, unless there's a consensus that `yt.utilities.decompose` isn't considered public API ?

Reply:
It's not public API, so this should be ok by me.

Reply:
oh! ya, I had assumed without too much thought that this was not public API. I think there is a small chance that there's an external frontend out there that utilizes this function. Happy to change this up to cover that scenario: the simplest change might be to have the return from this function vary and include a deprecation, e.g., 

```python
if cell_widths is not None:
    return left_edges, right_edges, shapes, slices, cell_widths_by_grid
else:
    deprecation warning
    return left_edges, right_edges, shapes, slices
```
Could leave out the deprecation but that might make future typing efforts annoying.

Or I could add a `split_array_with_cell_widths` function and just call that directly from the `load_uniform_grid` when needed. That'd probably result in a bit of code duplication, but it might be clearer to have a separate function. 



Also happy to leave this change as-is.

Reply:
Following John's remark I think the best course of action is to keep it simple and treat it as private. We can always inject additional backward-compatibility fixes later *if* anyone asks for it.